source/server/listener_manager_impl.cc

```diff
@@ -109,25 +109,38 @@ ProdListenerComponentFactory::createDrainManager(envoy::api::v2::Listener::Drain
 // This same object can be extended to handle additional listener socket options.
 class ListenerSocketOption : public Network::Socket::Option, Logger::Loggable<Logger::Id::config> {
 public:
-  ListenerSocketOption(bool transparent) : transparent_(transparent) {}
+  ListenerSocketOption(const envoy::api::v2::Listener& config)
+      : transparent_(config.transparent()) {}
 
   // Network::Socket::Option
-  bool setOption(Network::Socket& socket, bool pre_bind) const override {
-    // IP_TRANSPARENT makes a difference only for the following bind(). Set it before the bind
-    // and ignore it after.
-    if (transparent_ && pre_bind) {
+  bool setOption(Network::Socket& socket, Network::Socket::SocketState state) const override {
+    // IP_TRANSPARENT has an effect on bind() (allowing bind() to non-local addresses), but
+    // also after bind(), controlling whether new connections with non-local addresses are
+    // redirected to the listening socket or not, so we need to set it in either case, as
+    // it is possible to, e.g., set it on a listener and then change the listener configuration
+    // to not set it.
+    // This has the side effect of setting the option twice after the socket is first created,
+    // both before and after bind().
+    UNREFERENCED_PARAMETER(state);
+    int transparent = transparent_;
+
 #ifdef SOL_IP
-      int on = 1;
-      if (setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &on, sizeof(on)) == -1) {
-        ENVOY_LOG(warn, "Setting IP_TRANSPARENT on listener socket failed: {}", strerror(errno));
-        throw EnvoyException("ListenSocketOption: Error setting IP_TRANSPARENT socket option");
-      }
+    int rc = setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &transparent, sizeof(transparent));
+    // If we are unable to set the option we still return success for the default value (0),
+    // as we never changed the option to begin with. This also prevents failures in all current
+    // test cases that do not set the transparent option.
+    if (rc == 0 || transparent == 0) {
+      return true;
+    }
+    ENVOY_LOG(warn, "Setting IP_TRANSPARENT to {} on listener socket failed: {}", transparent,
+              strerror(errno));
 #else
-      UNREFERENCED_PARAMETER(socket);
-      throw EnvoyException("ListenSocketOption: Error setting IP_TRANSPARENT socket option");
-#endif
+    if (transparent == 0) {
+      return true;
     }
-    return true;
+#endif
+    throw EnvoyException("ListenSocketOption: Error setting IP_TRANSPARENT socket option");
+    return false;
```

Comment:
This line is unreachable. But shouldn't this just return false instead of throwing? I thought that was the convention for this function.

Reply:
The caller of this callback will also `throw()` in case of `false` return. I added a more specific `throw` here to be able to test that this implementation is called in the test case.

I was concerned that compiler would complain about not returning, but I was wrong, leaving the return statement out works as well, so I removed it.

Reply:
But isn't the API contract for this interface that it will return false on failure? If it is allowed to throw, I think the interface should return void. Or at a minimum, the documentation should specify in what cases it will throw vs when it could return false.